# Assignment 4.1

> Replace all TODOs with your code.
>
> Do not change any other code and do not add/remove cells!

## Sets

### Task 1

Create an empty set called 'my_set'.

In [1]:
my_set = set()

print(my_set)  # should print set()

set()


### Task 2

Let's turn a given sentence into a set of words.


In [2]:
sentence = "abc cba acd abc"

words =  set(sentence.split())

print(words)  # should be {'acd', 'cba', 'abc'} or in other order

{'acd', 'cba', 'abc'}


### Task 3

Now, let's remove from `words` all the words that are present in `sentence2`.

In [3]:
sentence2 = 'cba baa acc'
sentence2 = "cba baa acc"
words = set(sentence.split())
words2 = set(sentence2.split())
words -= words2

print(words)  # should print {'acd', 'abc'} or {'abc', 'acd'}

{'acd', 'abc'}


### Task 4

Let's now play around with statistics and probability!

Generate `N` random numbers in the range `[1, N]` and keep track of a percent of unique numbers in the result.
Play around with different values of `N`; how does it affect the percentage of unique numbers? How do you think why?

In [4]:
from random import randint

N = 10000000
numbers = [randint(1, N) for _ in range(N)]
unique_numbers = set(numbers)
percentage_unique = (len(unique_numbers) / N) * 100
print(f"N = {N}, Unique percentage: {percentage_unique:.2f}%")


N = 10000000, Unique percentage: 63.21%


For math geeks: if you set rather big `N`, you see that the percentage is slowly approaching ~63.21. It is a modification of a [Birthday problem](https://en.wikipedia.org/wiki/Birthday_problem), but a simpler and more relevant explanation is [here](https://math.stackexchange.com/questions/5775/how-many-bins-do-random-numbers-fill).


## Dictionaries

### Task 5

You are given a matrix `m`. Return a distinct list containing all the values that are present in at least `n` rows.

Hint: there can be duplicates in rows; perhaps we would want to eliminate them - how?

In [6]:
m = [
    [1, 1, 2, 3],
    [1, 2, 6, 7],
    [1, 7, 9, 9]
]

n = 2
res = []

count = {}

for row in m:
    unique_values = set(row)
    for value in unique_values:
        count[value] = count.get(value, 0) + 1

res = [key for key, val in count.items() if val >= n]

print(res)  # should print [1, 2, 7]

[1, 2, 7]
